### ISSUE TITLE
csi configmap should be updated by single controllers

### ISSUE THREAD

**Issue by Madhu-1** [rook/rook#13978]
Currently, the ceph reconciler subvolumegroup reconciler and radosnamespace reconciler updates/creates the csi configmap and this is error prone and easy to introduce bug in this area. Only the csi controller should create/update the configmap based on the resources created.

**Comment by parth-gr** [comment]
so we should have a watcher on csi controller to reconcile and update csi cm if any of the `subvolumegroup` `radosnamespace` is created/registered/updated?

**Comment by Madhu-1** [comment]
> so we should have a watcher on csi controller to reconcile and update csi cm if any of the `subvolumegroup` `radosnamespace` is created/deleted`

yes also we need to remove csi configmap create/update from the ceph cluster reconciler as well and ensure that we don't break anything

**Comment by parth-gr** [comment]
Sure, thanks for the feedback,
I would like to work on this /assign

**Comment by parth-gr** [comment]
Like to discuss for a better understanding the error-prone area in the code,
I see mostly the update functions of csi-config has the mutex lock with them in this file https://github.com/rook/rook/blob/master/pkg/operator/ceph/csi/cluster_config.go#L350

So they can handle the concurrent execution... 
So the ask is to better simply the calling of csi-cm at the same place or it help with other benefits, cc @Madhu-1

**Comment by BlaineEXE** [comment]
I also noticed this with my recent multus holder work, and I agree that it is something that could be error prone if/when we rearchitect the CSI controller. 

AFAIU, the `metadata.generation` field on the ConfigMap means that k8s will be able to detect race conditions, at which point one of the controllers will fail and have to reconcile again. However, I haven't tested this to be certain, and it does still seem good to me to keep CSI code as bug-free as possible.

I'm not sure the best way to resolve this, but I have one thought for consideration. 

Effectively, we want to limit write access to the primary CSI configmap. Only the CSI controller should be able to write to it. If other controllers (or users) need to write CSI config, they can have a separate, dedicated configmap that they are allowed to write to.

Let's call the primary configmap CM-A. Only the CSI controller is allowed to write to this CM. The second CM let's call CM-B. CM-B cannot be written by the CSI controller -- only read. Rook would write to CM-B to specify configs for SVGs or RadosNSs. The SVG and RadosNS controllers would use a mutex lock to make sure they don't try to write to CM-B simultaneously, as is used today.

I haven't thought deeply about this design, so there may be issues. For example, it relies on CM-B to be controlled by a single entity, which would limit user modifications. If user mods are a concern, or if multiple operators need to write configs (e.g., rook and provider-consumer operator), maybe CSI needs to be able to read from multiple configmaps. Alternately, we could rely on k8s's internal race condition checks, assuming those work as expected.

**Comment by Madhu-1** [comment]
@BlaineEXE what if only the csi controller watches for the  `CephCluster, Subvolumegroup,RadosNamespace` CRs and updates the configmap as required?

**Comment by parth-gr** [comment]
@Madhu-1 waiting on your final comments.
How do you see the overall design?

**Comment by Madhu-1** [comment]
> I think the CSI controller can watch whatever resource kinds it wants/needs to. That is certainly an option. The CSI team handles the largest part of the controller design and implementation, and my stance is that I don't want to force any design decisions onto it that aren't critical.
> 

i would like to have csi controller watching for the required resources and updating the configmap

> My larger consideration is this: We have had a few conversations recently about making the CSI controller into a separate library that can be consumed by other operators if desired. Whatever solution the CSI team works toward, my more future-looking thought is to make sure it's something that doesn't make that work harder, if we decide to do that work.

Yes, correct its currently in progress, yes most of the management logic will be moved to the csi (new operator) but still in think we need to have shim csi controller in rook to reconcile it.

Lets keep this on hold until we get some progress on the new csi work.

**Comment by parth-gr** [comment]
Should we close this in respect of https://github.com/rook/rook/pull/14089

**Comment by Madhu-1** [comment]
> Should we close this in respect of #14089 @Madhu-1 @BlaineEXE

not yet, we have worked on only the creation part in the above PR

**Comment by parth-gr** [comment]
@Madhu-1 I think this would be still a valid feature once integration is done.

**Comment by Madhu-1** [comment]
no its not, we can close this one once we move to the csi-operator
